### PR TITLE
Ingestor key interface

### LINKED ISSUE #70: task: [ingestor] create Key interface for ingestor
Create an Key interface that will be used to implement various key providers

### PR DESCRIPTION
**Pull request by pxp928**
Signed-off-by: pxp928 <parth.psu@gmail.com>

Adds key interface to retrieve, store and delete keys from the various key providers.

Closes #70

### CODE REVIEW

File: pkg/ingestor/key/key.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
+//
```
Comment (mlieberman85) [inline pkg/ingestor/key/key.go:1]: Do we want to store any hints about the key types at all, or would the expectation be that it is always just a collection of bytes and we have to figure out what key type it is?

**Comment by pxp928** [comment]
Moved key wrapper into the Key package such that it does not need to be done at the verifier package.

File: internal/testing/ingestor/keyutil/keyutil.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"testing"
+
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+func GetECDSAPubKey(t *testing.T) crypto.PublicKey {
```
Comment (lumjjb) [inline internal/testing/ingestor/keyutil/keyutil.go:30]: change this to not take in testing, and return err, and have test that calls this output err?

File: pkg/ingestor/key/key.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/rekor/pkg/pki/x509"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	rsaKeyType            string = "rsa"
```
Comment (lumjjb) [inline pkg/ingestor/key/key.go:35]: create alias type, `type KeyType string`

File: pkg/ingestor/key/key.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	StoreKey(id string, pk crypto.PublicKey) error
+	// DeleteKey takes in the ID and will remove the associated key from the provider
+	DeleteKey(id string) error
+	// Type returns the key provider type
+	Type() KeyProviderType
+}
+
+type Key struct {
+	// KeyHash sha256 hash of the canonical representation of the key
+	KeyHash string
```
Comment (lumjjb) [inline pkg/ingestor/key/key.go:61]: nit: remove Key prefix for attributes.

File: pkg/ingestor/key/key.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+func RegisterKeyProvider(k KeyProvider, providerType KeyProviderType) {
+	if _, ok := keyProviders[providerType]; ok {
+		logrus.Warnf("the key provider is being overwritten: %s", providerType)
+	}
+	keyProviders[providerType] = k
+}
+
+// Find goes through each of the registered key providers and retrieves the key
+// TODO: Should this handle if multiple keys are returned
+func Find(id string) (*Key, error) {
```
Comment (lumjjb) [inline pkg/ingestor/key/key.go:86]: Find and Retrieve have quite a bit of re-used code, can we refactor for find to use retrieve?

File: pkg/ingestor/key/key_test.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+			err := Store(tt.args.id, tt.args.pk, tt.args.providerType)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Store() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if err != nil {
+				if !strings.Contains(err.Error(), tt.wantErrMessage) {
+					t.Errorf("Retrieve() error = %s, wantErrMessage %s", err, tt.wantErrMessage)
+				}
+			}
+			if err == nil {
```
Comment (lumjjb) [inline pkg/ingestor/key/key_test.go:222]: nit: else {

**Review by lumjjb — APPROVED** [review]

File: pkg/ingestor/key/key.go (hunk 5 above)
Comment (pxp928) [reply pkg/ingestor/key/key.go:86]: ah yes good catch.

File: pkg/ingestor/key/key.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+}
+
+func getKeyInfo(pub crypto.PublicKey) (KeyType, KeyScheme, error) {
+	switch pub.(type) {
+	case *rsa.PublicKey:
+		return rsaKeyType, rsassapsssha256Scheme, nil
+	case *ecdsa.PublicKey:
+		return ecdsaKeyType, ecdsaSha2nistp256, nil
+	case ed25519.PublicKey:
```
Comment (mlieberman85) [inline pkg/ingestor/key/key.go:195]: The others are pointers to the key, this is just the key. Is that right?

**Review by mlieberman85 — APPROVED** [review]
LGTM

**Pull request merged**